effect/shared/src/main/scala/scalaz/effect/IORef.scala

```diff
@@ -15,28 +18,104 @@ package scalaz.effect
  * } yield ()
  * }}}
  */
-final class IORef[A] private (@volatile private var value: A) {
-
+final class IORef[A] private (private val value : AtomicReference[A]) extends AnyVal {
   /**
    * Reads the value from the `IORef`.
    */
-  final def read: IO[A] = IO.sync(value)
+  final def read[E]: IO[E, A] = IO.sync(value.get)
+
+  /**
+   * Writes a new value to the `IORef`, with a guarantee of immediate
+   * consistency (at some cost to performance).
+   */
+  final def write[E](a: A): IO[E, Unit] = IO.sync(value.set(a))
+
+  /**
+   * Writes a new value to the `IORef` without providing a guarantee of
+   * immediate consistency.
+   */
+  final def writeLater[E](a: A): IO[E, Unit] = IO.sync(value.lazySet(a))
+
+  /**
+   * Attempts to write a new value to the `IORef`, but aborts immediately under
+   * concurrent modification of the value by other fibers.
+   */
+  final def tryWrite[E](a: A): IO[E, Boolean] = IO.sync(value.compareAndSet(value.get, a))
 
   /**
-   * Writes a new value to the `IORef`.
+   * Atomically modifies the `IORef` with the specified function. This is not
+   * implemented in terms of `modifyFold` purely for performance reasons.
    */
-  final def write(a: A): IO[Unit] = IO.sync(this.value = a)
+  final def modify[E](f: A => A): IO[E, A] = IO.sync {
+    var loop = true
+    var next : A = null.asInstanceOf[A]
```

Comment:
😱

Reply:
Hehe. The other alternative is to `@tailcall` it and inspect the generated bytecode.

Unfortunately any unnecessary boxing in `IORef` will significantly decrease performance given how low-level the construct is...

Reply:
I hear you and trust the motivation 😄